src/request/export/ExportWords.js

```diff
@@ -21,94 +22,112 @@ class ExportWords extends Nimiq.Observable {
+        new ProgressIndicator($validateWordsPage.querySelector('.progress-indicator'), 3, 3);
+        /* eslint-enable no-new */
+
+        // events
+        $noRecoverySkip.addEventListener('click', event => {
+            event.preventDefault();
+            this._resolve({ success: false });
+        });
+        $recoveryWordsSkip.addEventListener('click', event => {
+            event.preventDefault();
+            this._resolve({ success: true });
+        });
+        this._wordsPasswordBox.on(PassphraseBox.Events.SUBMIT, this._passphraseSubmitted.bind(this));
+        $recoveryWordsContinue.addEventListener('click', () => {
+            this._validateWords.reset();
+            window.location.hash = ExportWords.Pages.VALIDATE_WORDS;
+        });
+        this._validateWords.on(ValidateWords.Events.VALIDATED, () => this._resolve({ success: true }));
+        this._validateWords.on(ValidateWords.Events.SKIP, () => this._resolve({ success: true }));
     }
 
     run() {
-        this._privacyWarningPassphraseBox.reset();
-        window.location.hash = ExportWords.Pages.PRIVACY;
+        this._wordsPasswordBox.reset();
+        window.location.hash = ExportWords.Pages.RECOVERY_WORDS_INTRO;
         if (TopLevelApi.getDocumentWidth() > Constants.MIN_WIDTH_FOR_AUTOFOCUS) {
-            this._privacyWarningPassphraseBox.focus();
+            this._wordsPasswordBox.focus();
         }
     }
 
     /**
-     * @param {string} phrase
+     * @param {string} password
      */
-    async _passphraseSubmitted(phrase) {
+    async _passphraseSubmitted(password) {
         TopLevelApi.setLoading(true);
-        const passphraseBuffer = phrase ? Utf8Tools.stringToUtf8ByteArray(phrase) : undefined;
+
+        let passwordBuffer;
+        if (this._password) {
+            passwordBuffer = this._password;
+        } else if (password) {
+            passwordBuffer = Utf8Tools.stringToUtf8ByteArray(password);
+        }
+
         /** @type {Key?} */
         let key = null;
+
         try {
-            key = await KeyStore.instance.get(this._request.keyInfo.id, passphraseBuffer);
+            key = await KeyStore.instance.get(this._request.keyInfo.id, passwordBuffer);
         } catch (e) {
             if (e.message === 'Invalid key') {
+                this._wordsPasswordBox.onPassphraseIncorrect();
                 TopLevelApi.setLoading(false);
-                this._privacyWarningPassphraseBox.onPassphraseIncorrect();
                 return;
             }
             this._reject(new Errors.CoreError(e));
             return;
         }
+
         if (!key) {
             this._reject(new Errors.KeyNotFoundError());
             return;
         }
 
-        this.setKey(key);
-        this.fire(ExportWords.Events.KEY_CHANGED, {
-            key,
-            isProtected: this._request.keyInfo.encrypted,
-        });
+        this.setKey(key, passwordBuffer);
+        this.fire(ExportWords.Events.KEY_CHANGED, key, passwordBuffer);
```

Comment:
Is it really necessary to fire both the key and the password? I think the key should suffice. At least from ExportFile to ExportWords, only the key is enough. When ExportFile then uses the database-stored encrypted secret, instead of decrypting and encrypting, the key (and password) would not have to be fired from ExportWords.

Reply:
The words do not need the password. The file does as it stands. I looked at Import words code for it where it still has the comment about having the key encrypted twice. I did not change this behaviour in this PR, but we can look into it. I wanted both events to fire the same parameters, but I can change it.